File: csharp/954.array-of-doubled-pairs.cs
/*
 * @lc app=leetcode id=954 lang=csharp
 *
 * [954] Array of Doubled Pairs
 */

using System;
using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {

    private static bool CanReorder(List<int> nums) {
        int n = nums.Count;
        if (n % 2 == 1) {
            return false;
        }
        nums.Sort();
        var map = nums
            .Select((int num, int index) => (Num: num, Index: index))
            .GroupBy(tuple => tuple.Num, (int num, IEnumerable<(int Num, int Index)> tuples) => (Num: num, Indices: tuples.Select(tuple => tuple.Index).ToList()))
            .ToDictionary(tuple => tuple.Num, tuple => tuple.Indices);
        for (int i = 0; i < n; ++i) {
            int curr = nums[i];
            var list = map[curr];
            if (list.Count == 0 || list[0] != i) {
                continue;
            }
            list.RemoveAt(0);
            if (!map.TryGetValue(2 * curr, out var next) || next.Count == 0) {
                return false;
            }
            next.RemoveAt(0);
        }
        return true;
    }

    public bool CanReorderDoubled(int[] A) {
        int zc = 0;
        List<int> l = new List<int>(), r = new List<int>();
        foreach (int x in A) {
            if (x < 0) {
                l.Add(-x);
            } else if (x == 0) {
                ++zc;
            } else {
                r.Add(x);
            }
        }
        return zc % 2 == 0 && CanReorder(l) && CanReorder(r);
    }
}
// @lc code=end

File: java/954.array-of-doubled-pairs.java
/*
 * @lc app=leetcode id=954 lang=java
 *
 * [954] Array of Doubled Pairs
 */

import java.util.List;
import java.util.TreeMap;

// @lc code=start
class Solution {
    public boolean canReorderDoubled(int[] arr) {
        TreeMap<Integer, Integer> pos = new TreeMap<>();
        TreeMap<Integer, Integer> neg = new TreeMap<>();
        int zeroCount = 0, posCount = 0, negCount = 0;
        for (int x : arr) {
            if (x == 0) {
                ++zeroCount;
            } else if (x > 0) {
                ++posCount;
                pos.put(x, pos.getOrDefault(x, 0) + 1);
            } else {
                ++negCount;
                neg.put(-x, neg.getOrDefault(-x, 0) + 1);
            }
        }
        if (zeroCount % 2 == 1 || posCount % 2 == 1 || negCount % 2 == 1) {
            return false;
        }
        return canReorderDoubled(pos) && canReorderDoubled(neg); 
    }

    private static boolean canReorderDoubled(TreeMap<Integer, Integer> vals) {
        List<Integer> keys = vals.keySet().stream().toList();
        for (int key : keys) {
            if (!vals.containsKey(key)) {
                continue;
            }
            int doubled = 2 * key;
            if (!vals.containsKey(doubled)) {
                return false;
            }
            int val = vals.get(key), doubleVal = vals.get(doubled);
            if (doubleVal < val) {
                return false;
            }
            vals.remove(key);
            if (doubleVal == val) {
                vals.remove(doubled);
            } else {
                vals.put(doubled, doubleVal - val);
            }
        }
        return true;
    }
}
// @lc code=end

